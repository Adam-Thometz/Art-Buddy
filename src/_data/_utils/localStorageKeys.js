// visited state for Word-To-Music Decoder
export const WTM = 'wtm';
// visited state for Jump Into Rhythm
export const JIR = 'jir';
// visited state for Sequence Maker
export const SM = 'sm';

// visited state for Free Paint
export const FP = 'fp';

// visited state for Score Keeper
export const SK = 'sk';
// visited state for Time Keeper
export const TK = 'tk';
// visited state for Mood Meter
export const MM = 'mm';