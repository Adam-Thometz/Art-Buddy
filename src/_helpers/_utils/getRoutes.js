/** getRoutes
 * Purpose: takes a React component and returns its children. Mainly used to get Route components found in a larger component and spread them in a Routes component.
 */

export default function getRoutes(component) {
  return component().props.children;
};