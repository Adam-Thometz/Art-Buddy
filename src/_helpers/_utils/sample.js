import { Sampler, Volume } from "tone";

/** sample
 * Purpose: takes a sound and a number for volume and returns a sample to play notes with. Used mainly in play-related function
 */

const sample = (sound, volume) => {
  const vol = new Volume(volume).toDestination();
  return new Sampler({
    urls: { C3: sound }
  }).connect(vol);
};

export default sample;