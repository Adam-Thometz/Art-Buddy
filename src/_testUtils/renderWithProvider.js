import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

// import renderWithRouter from './renderWithRouter';
import rootReducer from '../_redux/reducers/rootReducer';

function withProvider({ children, store }) {
  return (
    <Provider store={store}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
};

function renderWithProvider(component, options) {
  const store = (options && options.store)
    ? options.store
    : rootReducer;
  return render(component, { wrapper: (args) =>
    withProvider({
      ...args,
      store
    }), ...options
  });
};

export default renderWithProvider;