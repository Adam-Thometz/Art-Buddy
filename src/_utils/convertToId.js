export default function convertToId(name) {
  const nameArr = name.split(' ');
  let id = '';
  id += nameArr[0].toLowerCase();
  for (let i = 1; i < nameArr.length; i++) {
    nameArr[i].toLowerCase();
    nameArr[i][0].toUpperCase();
    id += nameArr[i];
  };
  return id;
};