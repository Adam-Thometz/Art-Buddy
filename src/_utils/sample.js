import { Sampler } from "tone";

const sample = (sound) => (
  new Sampler({
    urls: {
      C3: sound
    }
  }).toDestination()
);

export default sample;