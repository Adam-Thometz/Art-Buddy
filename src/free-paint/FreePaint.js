import React, { useEffect } from 'react';
import useVisited from '_hooks/useVisited';

import { useSelector, useDispatch } from 'react-redux';
import { clearGame } from '_redux/free-paint/freePaintActions';
import { changeCurrGame } from '_redux/_general/generalActions';

import WindowNavbar from '_components/window-nav/WindowNavbar';
import Instructions from '_components/instructions/Instructions';
import ControlBar from './control-bar/ControlBar';
import Canvas from './canvas/Canvas';
import Stencil from './stencil/Stencil';

import { freePaint } from '_data/_activities/activityList';

const FreePaint = () => {
  const [hasVisited] = useVisited(freePaint.lsKey);
  const { currGame } = useSelector(state => state.general);
  const { stencil } = useSelector(state => state.freePaint);
  const dispatch = useDispatch();
  
  useEffect(() => {
    dispatch(changeCurrGame(freePaint));
    return () => {
      dispatch(clearGame())
      dispatch(changeCurrGame({}));
    };
  }, [dispatch]);

  return (<>
    <WindowNavbar page={currGame.name} />
    {!hasVisited ? <Instructions game={currGame} /> : <>
      <ControlBar />
      <Canvas />
      {stencil ? <Stencil /> : null}
    </>}
  </>);
};

export default FreePaint;