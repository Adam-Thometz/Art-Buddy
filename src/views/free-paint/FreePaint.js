import { useEffect } from "react";
import useReportCard from "hooks/report-card/useReportCard";
import useVisited from "hooks/visited/useVisited";

import { useSelector, useDispatch } from "react-redux";
import { clearGame, setAvailableColors } from "store/free-paint/freePaintReducer";
import { changeCurrGame } from "store/_general/generalReducer";

import WindowNavbar from "components/window-nav/WindowNavbar";
import Instructions from "components/instructions/Instructions";
import ControlBar from "./control-bar/ControlBar";
import Canvas from "./canvas/Canvas";
import Stencil from "./stencil/Stencil";

import { freePaint } from "data/_activities/activityList";
import colorWheel from "data/color-theory/colorWheel";
import checkHasPassed from "utils/_report-card/checkHasPassed";

const FreePaint = () => {
  const [hasVisited, setHasVisited] = useVisited(freePaint.lsKey);
  const { currGame } = useSelector((state) => state.general);
  const { stencil } = useSelector((state) => state.freePaint);
  const [reportCard] = useReportCard("colorTheory");
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(changeCurrGame(freePaint));
    return () => {
      dispatch(clearGame());
      dispatch(changeCurrGame({}));
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    const availableColors = Object.keys(colorWheel);
    const colorsToAdd = [];
    if (colorsToAdd.length + 3 !== availableColors.length) {
      let jumpRate = 0;
      if (checkHasPassed(reportCard, ["tertiaryA", "tertiaryB"])) {
        jumpRate = 1;
      } else if (checkHasPassed(reportCard, ["secondaryA", "secondaryB"])) {
        jumpRate = 2;
      } else if (checkHasPassed(reportCard, ["primaryA", "primaryB"])) {
        jumpRate = 4;
      }

      if (jumpRate) {
        for (let i = 0; i < availableColors.length; i += jumpRate) {
          colorsToAdd.push(colorWheel[availableColors[i]].color.hex);
        }
        dispatch(setAvailableColors(colorsToAdd));
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <WindowNavbar page={currGame.name} />
      {!hasVisited ? (
        <Instructions setVisited={setHasVisited} />
      ) : (
        <>
          <ControlBar />
          <Canvas />
          {stencil ? <Stencil /> : null}
        </>
      )}
    </>
  );
};

export default FreePaint;
