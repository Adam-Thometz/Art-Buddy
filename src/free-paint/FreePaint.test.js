import React from "react";

import renderWithProvider from '_testUtils/renderWithProvider';
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import FreePaint from './FreePaint';

import { FP } from '_data/_utils/localStorageKeys';

describe('FreePaint component', () => {
  window.localStorage.setItem(`visited-${FP}`, true);
  it('renders without crashing', () => {
    renderWithProvider(<FreePaint />);
  });
  
  it('matches snapshot', () => {
    const { asFragment } = renderWithProvider(<FreePaint />);
    expect(asFragment()).toMatchSnapshot();
  });
  
  it('renders a stencil when an option is picked', () => {
    renderWithProvider(<FreePaint />);
    const abc = screen.getByText('abc');
    userEvent.click(abc);
    const f = screen.getByText('f');
    userEvent.click(f);
    userEvent.click(abc);
    expect(screen.getByText('f')).toBeInTheDocument();
  });
});