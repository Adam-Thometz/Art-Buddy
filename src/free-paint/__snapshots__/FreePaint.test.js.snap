// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreePaint component matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="WindowNavbar-games"
  >
    <h1
      class="WindowNavbar-label"
    >
      FREE PAINT
    </h1>
    <img
      alt="Help"
      class="Help"
      src="file"
    />
  </header>
  <section
    class="ControlBar"
  >
    <div
      class="ControlBar-button-wrapper"
    >
      <button
        class="Button regular"
        id="upperCase"
        style="border-color: #17255a;"
      >
        ABC
      </button>
    </div>
    <div
      class="ControlBar-button-wrapper"
    >
      <button
        class="Button regular"
        id="lowerCase"
        style="border-color: #89eb0a;"
      >
        abc
      </button>
    </div>
    <div
      class="ControlBar-button-wrapper"
    >
      <button
        class="Button regular"
        id="numbers"
        style="border-color: #ffe81a;"
      >
        123
      </button>
    </div>
    <div
      class="ControlBar-button-wrapper"
    >
      <button
        class="Button regular"
        id="shapes"
        style="border-color: #fd4d3a;"
      >
        <img
          alt="Shapes"
          class="ControlBar-shape-button"
          src="file"
        />
      </button>
    </div>
    <div
      class="ControlBar-button-wrapper"
    >
      <button
        class="Button regular"
        id="colors"
        style="border-color: #000000; color: rgb(0, 0, 0);"
      >
        <figure
          class="Icon"
          id="colors"
        >
          <img
            alt=""
            src="file"
            style="height: 45px; width: 45px; cursor: pointer; filter: invert(0%) sepia(0%) saturate(0%) hue-rotate(304deg) brightness(100%) contrast(102%);"
          />
          <figcaption
            class="Icon-text"
            style="font-size: 1rem;"
          >
            Pencil
          </figcaption>
        </figure>
      </button>
    </div>
    <div
      class="ControlBar-button-wrapper"
    >
      <button
        class="Button regular"
      >
        <figure
          class="Icon"
        >
          <img
            alt=""
            src="file"
            style="height: 45px; width: 45px;"
          />
          <figcaption
            class="Icon-text"
            style="font-size: 1rem;"
          >
            Eraser
          </figcaption>
        </figure>
      </button>
    </div>
  </section>
  <canvas
    class="Canvas"
    height="0"
    style="width: 85%; height: 85%;"
    width="0"
  />
  <div
    class="Canvas-bottom-options"
  >
    <span>
      clear canvas
    </span>
    <span>
      clear stencil
    </span>
  </div>
</DocumentFragment>
`;
